timeboost/bid_validator.go

```diff
@@ -0,0 +1,372 @@
+		redisClient:               redisClient,
+		stack:                     stack,
+		auctionContract:           auctionContract,
+		auctionContractAddr:       auctionContractAddr,
+		bidsReceiver:              make(chan *Bid, 10_000),
+		initialRoundTimestamp:     initialTimestamp,
+		roundDuration:             roundDuration,
+		auctionClosingDuration:    auctionClosingDuration,
+		reserveSubmissionDuration: reserveSubmissionDuration,
+		reservePrice:              reservePrice,
+		domainValue:               domainValue,
+		bidsPerSenderInRound:      make(map[common.Address]uint8),
+		maxBidsPerSenderInRound:   5, // 5 max bids per sender address in a round.
+		producerCfg:               &cfg.ProducerConfig,
+	}
+	api := &BidValidatorAPI{bidValidator}
+	valAPIs := []rpc.API{{
+		Namespace: AuctioneerNamespace,
+		Version:   "1.0",
+		Service:   api,
+		Public:    true,
+	}}
+	stack.RegisterAPIs(valAPIs)
+	return bidValidator, nil
+}
+
+func EnsureBidValidatorExposedViaRPC(stackConf *node.Config) {
+	found := false
+	for _, module := range stackConf.HTTPModules {
+		if module == AuctioneerNamespace {
+			found = true
+			break
+		}
+	}
+	if !found {
+		stackConf.HTTPModules = append(stackConf.HTTPModules, AuctioneerNamespace)
+	}
+}
+
+func (bv *BidValidator) Initialize(ctx context.Context) error {
+	if err := pubsub.CreateStream(
+		ctx,
+		validatedBidsRedisStream,
+		bv.redisClient,
+	); err != nil {
+		return fmt.Errorf("creating redis stream: %w", err)
+	}
+	p, err := pubsub.NewProducer[*JsonValidatedBid, error](
+		bv.redisClient, validatedBidsRedisStream, bv.producerCfg,
+	)
+	if err != nil {
+		return fmt.Errorf("failed to init redis in bid validator: %w", err)
+	}
+	bv.producer = p
+	return nil
+}
+
+func (bv *BidValidator) Start(ctx_in context.Context) {
+	bv.StopWaiter.Start(ctx_in, bv)
+	if bv.producer == nil {
+		log.Crit("Bid validator not yet initialized by calling Initialize(ctx)")
+	}
+	bv.producer.Start(ctx_in)
+
+	// Thread to set reserve price and clear per-round map of bid count per account.
+	bv.StopWaiter.LaunchThread(func(ctx context.Context) {
+		reservePriceTicker := newAuctionCloseTicker(bv.roundDuration, bv.auctionClosingDuration+bv.reserveSubmissionDuration)
+		go reservePriceTicker.start()
+		clearBidCountTicker := newAuctionCloseTicker(bv.roundDuration, bv.auctionClosingDuration)
+		go clearBidCountTicker.start()
+
+		for {
+			select {
+			case <-ctx.Done():
+				log.Error("Context closed, autonomous auctioneer shutting down")
+				return
+			case <-reservePriceTicker.c:
+				rp, err := bv.auctionContract.ReservePrice(&bind.CallOpts{})
+				if err != nil {
+					log.Error("Could not get reserve price", "error", err)
+					continue
+				}
+
+				currentReservePrice := bv.fetchReservePrice()
+				if currentReservePrice.Cmp(rp) == 0 {
+					continue
+				}
+
+				log.Info("Reserve price updated", "old", currentReservePrice.String(), "new", rp.String())
+				bv.setReservePrice(rp)
+
+			case <-reservePriceTicker.c:
```

Comment:
Should this be clearBidCountTicker?
```suggestion
			case <-clearBidCountTicker.c:
```

Reply:
Fixed it in https://github.com/OffchainLabs/nitro/pull/2561/commits/0d607f758c5de8c20136a0c547cd722d4c213679